From 7dd86345f2826f8c3d885474f3c21026f3f84ba6 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Thu, 11 May 2023 12:25:45 +0200
Subject: [PATCH 0395/1695] drivers: stm32_rng: add support for RNG power
 optimisation

Adds support for RNG V4.1 and above. These versions add a power
optimization and a modification of the seed error concealment.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 9f1e1c1e546e570d38ff02d6c7bf23af0e36132d drivers: stm32_rng: add support for RNG power optimisation]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I7227959484bf29846eb17eec057a24a84ccace2f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/304117
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_rng.c | 59 ++++++++++++++++++++++++++++++++++++----
 1 file changed, 53 insertions(+), 6 deletions(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index c9a4f7914..43133b024 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -34,6 +34,7 @@
 #define RNG_CR_CED		BIT(5)
 #define RNG_CR_CONFIG1		GENMASK_32(11, 8)
 #define RNG_CR_NISTC		BIT(12)
+#define RNG_CR_POWER_OPTIM	BIT(13)
 #define RNG_CR_CONFIG2		GENMASK_32(15, 13)
 #define RNG_CR_CLKDIV		GENMASK_32(19, 16)
 #define RNG_CR_CLKDIV_SHIFT	U(16)
@@ -66,6 +67,7 @@
 #define RNG_MAX_NOISE_CLK_FREQ	U(3000000)
 
 struct stm32_rng_driver_data {
+	bool has_power_optim;
 	bool has_cond_reset;
 	bool entropy_src_config;
 };
@@ -459,11 +461,35 @@ static TEE_Result stm32_rng_pm_resume(void)
 
 static TEE_Result stm32_rng_pm_suspend(void)
 {
-	vaddr_t base = get_base();
+	vaddr_t rng_base = get_base();
+
+	stm32_rng->pm_cr = io_read32(rng_base + RNG_CR);
+
+	if (stm32_rng->ddata->has_cond_reset) {
+		stm32_rng->pm_health = io_read32(rng_base + RNG_HTCR);
+		stm32_rng->pm_noise_ctrl = io_read32(rng_base + RNG_NSCR);
+	}
 
-	stm32_rng->pm_cr = io_read32(base + RNG_CR);
-	stm32_rng->pm_health = io_read32(base + RNG_HTCR);
-	stm32_rng->pm_noise_ctrl = io_read32(base + RNG_NSCR);
+	if (stm32_rng->ddata->has_power_optim) {
+		uint64_t timeout_ref = 0;
+
+		/*
+		 * As per reference manual, it is recommended to set
+		 * RNG_CONFIG2[bit0] when RNG power consumption is critical.
+		 */
+		io_setbits32(rng_base + RNG_CR, RNG_CR_POWER_OPTIM |
+				RNG_CR_CONDRST);
+		io_clrbits32(rng_base + RNG_CR, RNG_CR_CONDRST);
+
+		timeout_ref = timeout_init_us(RNG_READY_TIMEOUT_US);
+		while ((io_read32(rng_base + RNG_CR) & RNG_CR_CONDRST))
+			if (timeout_elapsed(timeout_ref))
+				break;
+		if ((io_read32(rng_base + RNG_CR) & RNG_CR_CONDRST))
+			panic();
+	} else {
+		io_clrbits32(rng_base + RNG_CR, RNG_CR_RNGEN);
+	}
 
 	return TEE_SUCCESS;
 }
@@ -594,6 +620,10 @@ static TEE_Result stm32_rng_probe(const void *fdt, int offs,
 	else
 		stm32mp_register_secure_periph_iomem(stm32_rng->base.pa);
 
+	/* Power management implementation expects both or none are set */
+	assert(stm32_rng->ddata->has_power_optim ==
+	       stm32_rng->ddata->has_cond_reset);
+
 	register_pm_core_service_cb(stm32_rng_pm, &stm32_rng, "rng-service");
 
 	return TEE_SUCCESS;
@@ -608,17 +638,34 @@ err:
 }
 
 static const struct stm32_rng_driver_data mp13_data[] = {
-	{ .has_cond_reset = true, .entropy_src_config = true },
+	{
+		.has_cond_reset = true,
+		.entropy_src_config = true,
+		.has_power_optim = true,
+	},
 };
 
 static const struct stm32_rng_driver_data mp15_data[] = {
-	{ .has_cond_reset = false, .entropy_src_config = false },
+	{
+		.has_cond_reset = false,
+		.entropy_src_config = false,
+		.has_power_optim = false,
+	},
 };
 DECLARE_KEEP_PAGER(mp15_data);
 
+static const struct stm32_rng_driver_data mp25_data[] = {
+	{
+		.has_cond_reset = true,
+		.entropy_src_config = true,
+		.has_power_optim = true,
+	},
+};
+
 static const struct dt_device_match rng_match_table[] = {
 	{ .compatible = "st,stm32-rng", .compat_data = &mp15_data },
 	{ .compatible = "st,stm32mp13-rng", .compat_data = &mp13_data },
+	{ .compatible = "st,stm32mp25-rng", .compat_data = &mp25_data },
 	{ }
 };
 
-- 
2.47.2

